src/vulkan_state.h

```diff
@@ -22,18 +22,42 @@
 
 #pragma once
 
+#include <functional>
 #include <vulkan/vulkan.hpp>
 
 #include "managed_resource.h"
+#include "vulkan_wsi.h"
 
-class VulkanWSI;
 
 class VulkanState
 {
 public:
-    VulkanState(VulkanWSI& vulkan_wsi);
+    using ChoosePhysicalDeviceStrategy = 
+        std::function<vk::PhysicalDevice (std::vector<vk::PhysicalDevice> const&)>;
+    
+    VulkanState(VulkanWSI& vulkan_wsi, ChoosePhysicalDeviceStrategy const& pd_strategy)
+    {
+        create_instance(vulkan_wsi);
+        create_physical_device(vulkan_wsi, pd_strategy);
+        create_logical_device(vulkan_wsi);
+        create_command_pool();
+    }
+
+    void create_physical_device(VulkanWSI& vulkan_wsi, ChoosePhysicalDeviceStrategy const& pd_strategy)
+    {
+        vk_physical_device = pd_strategy(available_devices(vulkan_wsi));
+    }
```

Comment:
All the above internal implementation details don't need to be exposed in the header. Plus create_physical_device() should be private.

Reply:
i dont know how it got to public sector :-) i wasn't thinking at all. And with exposing it i was probably just too lazy...